File: solutions/elixir/variable-length-quantity/2/lib/variable_length_quantity.ex
defmodule VariableLengthQuantity do
  import Bitwise

  @doc """
  Encode integers into a bitstring of VLQ encoded bytes
  """
  @spec encode(integers :: [integer]) :: binary
  def encode(integers) do
    integers
    |> Enum.map(&do_encode(&1, <<>>, 0))
    |> Enum.reduce(<<>>, &(&2 <> &1))
  end

  defp do_encode(0, <<>>, _), do: <<0>>
  defp do_encode(0, acc, _), do: acc

  defp do_encode(num, acc, sign) do
    do_encode(num >>> 7, <<sign::1, num::7>> <> acc, 1)
  end

  @doc """
  Decode a bitstring of VLQ encoded bytes into a series of integers
  """
  @spec decode(bytes :: binary) :: {:ok, [integer]} | {:error, String.t()}
  def decode(bytes) do
    do_decode(bytes, [], 0)
  end

  defp do_decode(<<0::1, num::7>>, acc, partial) do
    result = [(partial <<< 7) + num | acc]
    {:ok, Enum.reverse(result)}
  end

  defp do_decode(<<1::1, _::7>>, _, _), do: {:error, "incomplete sequence"}

  defp do_decode(<<0::1, num::7, bytes::binary>>, acc, partial) do
    do_decode(bytes, [(partial <<< 7) + num | acc], 0)
  end

  defp do_decode(<<1::1, num::7, bytes::binary>>, acc, partial) do
    do_decode(bytes, acc, (partial <<< 7) + num)
  end
end

File: solutions/typescript/variable-length-quantity/2/variable-length-quantity.ts
export function encode(input: number[]): number[] {
  return input.flatMap(encodeInt);
}

function encodeInt(num: number) {
  const chunks: number[] = [];

  do {
    const byte = num & 0x7f;
    const cont = chunks.length ? 0x80 : 0x0;
    chunks.push(byte | cont);
    num >>>= 7;
  } while (num > 0);

  return chunks.reverse();
}

export function decode(input: number[]) {
  const output: number[] = [];

  let current = 0;
  let partial = false;

  for (let byte of input) {
    current |= byte & 0x7f;
    partial = (byte & 0x80) != 0;

    if (!partial) {
      output.push(current >>> 0); // Ensure unsigned 32-bit
      current = 0;
    }

    current <<= 7;
  }

  if (partial) throw new Error("Incomplete sequence");

  return output;
}
